libraries/AP_Proximity/AP_Proximity_Boundary_3D.cpp

```diff
@@ -0,0 +1,136 @@
+#include "AP_Proximity_Backend.h"
+#include "AP_Proximity_Boundary_3D.h"
+
+/*
+  Constructor. 
+  This incorporates initialisation as well.
+*/
+AP_Proximity_Boundary_3D::AP_Proximity_Boundary_3D() 
+{
+    // initialise sector edge vector used for building the boundary fence
+    init_boundary();
+}
+
+// initialise the boundary and sector_edge_vector array used for object avoidance
+//   should be called if the sector_middle_deg or _sector_width_deg arrays are changed
+void AP_Proximity_Boundary_3D::init_boundary()
+{
+    for (uint8_t stack = 0; stack < PROXIMITY_NUM_STACK; stack ++) {
+        for (uint8_t sector=0; sector < PROXIMITY_NUM_SECTORS; sector++) {
+            float angle_rad = ((float)_sector_middle_deg[sector]+(PROXIMITY_SECTOR_WIDTH_DEG/2.0f));
+            float pitch = ((float)_pitch_middle_deg[stack]);
+            _sector_edge_vector[sector][stack].offset_bearing(angle_rad, pitch, 100.0f);
+            _boundary_points[sector][stack] = _sector_edge_vector[sector][stack] * PROXIMITY_BOUNDARY_DIST_DEFAULT;
+        }
+    }
+}
+
+// update boundary points used for object avoidance based on a single sector and pitch distance changing
+//   the boundary points lie on the line between sectors meaning two boundary points may be updated based on a single sector's distance changing
+//   the boundary point is set to the shortest distance found in the two adjacent sectors, this is a conservative boundary around the vehicle
+void AP_Proximity_Boundary_3D::update_boundary(const uint8_t sector, const uint8_t stack) {
```

Comment:
as mentioned, let's replace "stack" with "layer".

Reply:
we do a lot of work in this function to ensure that we don't let the vehicle hit the corner of an object.  I suspect that we are only being conservative in the horizontal direction meaning we can probably clip the corner vertically?
![image](https://user-images.githubusercontent.com/1498098/99203539-3f28ab00-27f6-11eb-8a48-f6c1d47eebb8.png)

Reply:
Yes you're right. When the object appears at a distance to some face, we pull in the adjacent faces also to be conservative. We could also pull in the layers and replicate a similar behaviour but I don't think its going to be very fruitful... Code complexity will definetly increase.